# Heroes Of Pymol Observations

-----

* Of the 780 active players, the vast majority are male (~84%). There also exists, a smaller, but notable proportion of female players (~14%).
* Our peak age demographic falls between 20-24 (46.79%) with secondary groups falling between 15-19 (17.44%) and 25-29 (12.95%). 
* Of the 780 purchases made, there were 183 unique items that sold for an average price of $\$$ 3.052 for a total revenue of $\$$ 2379.77


* Breaking down the purchases
    * by gender showed:
        * That 113 of the 780 purchases were made by females for an average price of $\$$ 3.20 for a total of $\$$ 361.94
        * That 652 of the 780 purchasese were made by males for an average price of $\$$ 3.02 for a total of $\$$ 1967.64
        * That 15 of the 780 purchasese were made by someone with other/non-disclosed gender for an average price of $\$$ 3.35 for a total of $\$$ 50.19
    * by age showed:
        * The 20-24 age group purchased the most items,	365 out of the 780 total purchases, for an average price of $\$$ 3.05, creating a total revenue of $\$$ 1,114.06.
        * The 15-19 age group purchased the second most items, 136 out of the 780 total purchases, for an average price of $\$$ 3.04, creating a total revenue of $\$$ 412.89.
        * The 25-29 age group purchased the third most items, 101 out of the 780 total purchases, for an average price of $\$$ 2.90, creating a total revenue of $\$$ 293.00.
        
        
* The top spenders:
    * The top spender was Lisosia93, making 5 purchases at an average of $\$$ 3.79 and generating a total of $\$$ 18.96.
    * Idastidru52 spent the next most money, making 4 purchases at an average of $\$$ 3.86 and generating a total of $\$$ 15.45.
    * Chamjask73 spent the third most amount amount of money, making 3 purchases at an average of $\$$4.61 and generating a total of $\$$ 13.83.


* The most popular item was by far Oathbreaker, Last Hope of the Breaking Storm, which sold 12 units. Following Oathbreaker, there was a 3-way tie for second place between Fiery Glass Crusader; Extraction, Quickblade Of Trembling Hands; and Nirvana, with each selling 9 units each.

* The most profitable item was Oathbreaker, Last Hope of the Breaking Storm, which brought in $\$$ 50.76, followed by Nirvana with $\$$ 44.10, and Fiery Glass Crusader with $\$$ 41.22.

-----

# Data Analysis

## Dependencies and Setup

In [1]:
import pandas as pd
import numpy as np

# Raw data file
file_to_load = "Resources/purchase_data.csv"

# Read purchasing file and store into pandas data frame
purchase_data = pd.read_csv(file_to_load)
purchase_data.head(15)
# output:
# 780 rows × 7 columns

Purchase ID             SN  Age                 Gender  Item ID  \
0             0        Lisim78   20                   Male      108   
1             1    Lisovynya38   40                   Male      143   
2             2     Ithergue48   24                   Male       92   
3             3  Chamassasya86   24                   Male      100   
4             4      Iskosia90   23                   Male      131   
5             5        Yalae81   22                   Male       81   
6             6      Itheria73   36                   Male      169   
7             7    Iskjaskst81   20                   Male      162   
8             8      Undjask33   22                   Male       21   
9             9    Chanosian48   35  Other / Non-Disclosed      136   
10           10      Inguron55   23                   Male       95   
11           11   Haisrisuir60   23                   Male      162   
12           12   Saelaephos52   21                   Male      116   
13           13    Assjaskan73   22                   Male        4   
14           14    Saesrideu94   35                   Male      165   

                                    Item Name  Price  
0   Extraction, Quickblade Of Trembling Hands   3.53  
1                           Frenzied Scimitar   1.56  
2                                Final Critic   4.88  
3                                 Blindscythe   3.27  
4                                        Fury   1.44  
5                                   Dreamkiss   3.61  
6      Interrogator, Blood Blade of the Queen   2.18  
7                               Abyssal Shard   2.67  
8                                   Souleater   1.10  
9                Ghastly Adamantite Protector   3.58  
10                      Singed Onyx Warscythe   4.74  
11                              Abyssal Shard   2.67  
12                    Renewed Skeletal Katana   4.18  
13                         Bloodlord's Fetish   1.70  
14                Bone Crushing Silver Skewer   4.86

## Player Count
* Display the total number of players

In [2]:
# extract the columns from the purchase_data dataframe that contain player information
player_info = purchase_data.loc[:, ["Gender", "SN", "Age"]]
player_info.head(10)
# output:
# 780 rows × 3 columns

Gender             SN  Age
0                   Male        Lisim78   20
1                   Male    Lisovynya38   40
2                   Male     Ithergue48   24
3                   Male  Chamassasya86   24
4                   Male      Iskosia90   23
5                   Male        Yalae81   22
6                   Male      Itheria73   36
7                   Male    Iskjaskst81   20
8                   Male      Undjask33   22
9  Other / Non-Disclosed    Chanosian48   35

In [3]:
# totalUniquePlayers = len(purchase_data["SN"].value_counts().dropna(how='any'))
# totalUniquePlayers = len(player_info["SN"].unique())
# most concise way:
totalUniquePlayers = player_info["SN"].nunique()

# insert into a new dataframe
totalUniquePlayers_df = pd.DataFrame({"Total Unique Players": [totalUniquePlayers]})
totalUniquePlayers_df

Total Unique Players
0                   576

## Gender Demographics
* Percentage and Count of Male Players
* Percentage and Count of Female Players
* Percentage and Count of Other / Non-Disclosed

### Unique player problem:
Using unique players yields a percentage greater than 100 % since some players are in the database multiple times, so we need to use total playersand not unique players in the database to get the actual gender breakdown.

In [4]:
#males = purchase_data.groupby("Gender").count()
#males = purchase_data["Gender"].count()
#print (males)
gender_breakdown_total = purchase_data["Gender"].value_counts()
print (gender_breakdown_total) # gender_breakdown_total is of type list
print("*"*50)

genders_df = pd.DataFrame({"Total Count": gender_breakdown_total})
genders_df["Percentage of total players"] = 100*(genders_df["Total Count"]/totalUniquePlayers)
genders_df

Male                     652
Female                   113
Other / Non-Disclosed     15
Name: Gender, dtype: int64
**************************************************


Total Count  Percentage of total players
Male                           652                   113.194444
Female                         113                    19.618056
Other / Non-Disclosed           15                     2.604167

In [5]:
# output greater than 100% so double checking
print (gender_breakdown_total/totalUniquePlayers)
print("*"*35)

print (f"Unique players: {totalUniquePlayers}")
print("*"*35)

totalPlayers = purchase_data["SN"].count()
print(f"Total players: {totalPlayers}")
print("*"*35)

genders_df["Percentage of unique players"] = 100*(gender_breakdown_total/totalPlayers)
genders_df

Male                     1.131944
Female                   0.196181
Other / Non-Disclosed    0.026042
Name: Gender, dtype: float64
***********************************
Unique players: 576
***********************************
Total players: 780
***********************************


Total Count  Percentage of total players  \
Male                           652                   113.194444   
Female                         113                    19.618056   
Other / Non-Disclosed           15                     2.604167   

                       Percentage of unique players  
Male                                      83.589744  
Female                                    14.487179  
Other / Non-Disclosed                      1.923077

### Proper way to find gender breakdown:
Since using the total number of players in a list that contains multiple entries for an individual player skews the final gender percentage, the most accurate way of obtaining the gender breakdown for the players in this database would be to drop duplicate entries of a player and recompute the gender percentage with each player appearing only once in the list.

In [6]:
players_list = player_info["SN"].value_counts()
players_list.head(15)
# output:
# Name: SN, Length: 576, dtype: int64

Lisosia93       5
Idastidru52     4
Iral74          4
Chadolyla44     3
Phaena87        3
Yathecal82      3
Pheodaisun84    3
Hada39          3
Aelin32         3
Iri67           3
Saistyphos30    3
Sondastsda82    3
Ialallo29       3
Chamimla85      3
Umolrian85      3
Name: SN, dtype: int64

In [7]:
unique_players_df = purchase_data.drop_duplicates(subset=['SN'], keep='first')
unique_players_df.head(10)
# output:
# 576 rows × 8 columns

Purchase ID             SN  Age                 Gender  Item ID  \
0            0        Lisim78   20                   Male      108   
1            1    Lisovynya38   40                   Male      143   
2            2     Ithergue48   24                   Male       92   
3            3  Chamassasya86   24                   Male      100   
4            4      Iskosia90   23                   Male      131   
5            5        Yalae81   22                   Male       81   
6            6      Itheria73   36                   Male      169   
7            7    Iskjaskst81   20                   Male      162   
8            8      Undjask33   22                   Male       21   
9            9    Chanosian48   35  Other / Non-Disclosed      136   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44  
5                                  Dreamkiss   3.61  
6     Interrogator, Blood Blade of the Queen   2.18  
7                              Abyssal Shard   2.67  
8                                  Souleater   1.10  
9               Ghastly Adamantite Protector   3.58

In [8]:
unique_players_list =  unique_players_df["SN"].value_counts()
unique_players_list.head(10)
# output: 
# Name: SN, Length: 576, dtype: int64

Assilsan72       1
Jiskirran77      1
Yasur35          1
Eolan54          1
Undassa89        1
Maluncil97       1
Eulanurin88      1
Chamadarnya73    1
Qilatista90      1
Sausosia74       1
Name: SN, dtype: int64

In [9]:
gender_breakdown_unique = unique_players_df["Gender"].value_counts()
print (gender_breakdown_unique)
print("*"*50)

unique_genders_df = pd.DataFrame(gender_breakdown_unique)
# genders_df
print (totalUniquePlayers)

unique_genders_df = pd.DataFrame({"Total Count": gender_breakdown_unique})
unique_genders_df["Percentage of total players"] = 100*(genders_df["Total Count"]/totalUniquePlayers)
unique_genders_df

unique_genders_df["Percentage of unique players"] = 100*(unique_genders_df["Total Count"]/totalUniquePlayers)
unique_genders_df

Male                     484
Female                    81
Other / Non-Disclosed     11
Name: Gender, dtype: int64
**************************************************
576


Total Count  Percentage of total players  \
Male                           484                   113.194444   
Female                          81                    19.618056   
Other / Non-Disclosed           11                     2.604167   

                       Percentage of unique players  
Male                                      84.027778  
Female                                    14.062500  
Other / Non-Disclosed                      1.909722

### Direct comparison of player gender breakdown before and after dropping duplicate entries:
As we can see, dropping duplicates from the original dataframe reduced the number of male players by 168 (from 652 down theo 484), reduced the number of female players by 32 (from 113 down to 81), and reduced the number of undisclosed genders by 4 (down from 15 to 11).

With such a small difference in total overall numbers, the gender breakdown changed minimally, from 83.59% to 84.02% for males, from 14.49% to 14.06% for females, and from 1.92% to 1.91% for undisclosed/other.

Since the overall percentage for the gender breakdown didn't change much regardless of whether or not we kept the duplicate players in the dataframe, dropping the duplicates in _this specific_ case wasn't necessary. However, this was a good practice in checking how the results varied depending on how we analyzed the data, and an indication that the results of any data analysis is contingent on how the data is interpreted.

In [10]:
genders_df[["Total Count", "Percentage of total players"]]

Total Count  Percentage of total players
Male                           652                   113.194444
Female                         113                    19.618056
Other / Non-Disclosed           15                     2.604167

In [11]:
unique_genders_df

Total Count  Percentage of total players  \
Male                           484                   113.194444   
Female                          81                    19.618056   
Other / Non-Disclosed           11                     2.604167   

                       Percentage of unique players  
Male                                      84.027778  
Female                                    14.062500  
Other / Non-Disclosed                      1.909722

## Age Demographics
* Establish bins for ages
* Categorize the existing players using the age bins using pd.cut()
* Calculate the numbers and percentages by age group
* Create a summary data frame to hold the results
* Round the percentage column to two decimal points
* Display Age Demographics Table

In [12]:
# Establish bins for ages
age_bins = [0, 9.90, 14.90, 19.90, 24.90, 29.90, 34.90, 39.90, 99999]
group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

# Categorize the existing players using the age bins using pd.cut()
purchase_data["Age Breakdown"] = pd.cut(purchase_data["Age"], age_bins, labels = group_names)
purchase_data.head(15)
# output: 780 rows × 8 columns

Purchase ID             SN  Age                 Gender  Item ID  \
0             0        Lisim78   20                   Male      108   
1             1    Lisovynya38   40                   Male      143   
2             2     Ithergue48   24                   Male       92   
3             3  Chamassasya86   24                   Male      100   
4             4      Iskosia90   23                   Male      131   
5             5        Yalae81   22                   Male       81   
6             6      Itheria73   36                   Male      169   
7             7    Iskjaskst81   20                   Male      162   
8             8      Undjask33   22                   Male       21   
9             9    Chanosian48   35  Other / Non-Disclosed      136   
10           10      Inguron55   23                   Male       95   
11           11   Haisrisuir60   23                   Male      162   
12           12   Saelaephos52   21                   Male      116   
13           13    Assjaskan73   22                   Male        4   
14           14    Saesrideu94   35                   Male      165   

                                    Item Name  Price Age Breakdown  
0   Extraction, Quickblade Of Trembling Hands   3.53         20-24  
1                           Frenzied Scimitar   1.56           40+  
2                                Final Critic   4.88         20-24  
3                                 Blindscythe   3.27         20-24  
4                                        Fury   1.44         20-24  
5                                   Dreamkiss   3.61         20-24  
6      Interrogator, Blood Blade of the Queen   2.18         35-39  
7                               Abyssal Shard   2.67         20-24  
8                                   Souleater   1.10         20-24  
9                Ghastly Adamantite Protector   3.58         35-39  
10                      Singed Onyx Warscythe   4.74         20-24  
11                              Abyssal Shard   2.67         20-24  
12                    Renewed Skeletal Katana   4.18         20-24  
13                         Bloodlord's Fetish   1.70         20-24  
14                Bone Crushing Silver Skewer   4.86         35-39

In [13]:
# Calculate the Numbers and Percentages by Age Group
groupedAges_totals = purchase_data["Age Breakdown"].value_counts(ascending=False)
groupedAges_percents = groupedAges_totals / totalPlayers * 100
groupedAges = pd.DataFrame({"Total Count": groupedAges_totals, "Percentage of Players": groupedAges_percents})
groupedAges

# Give the displayed data cleaner formatting by rounding dollar amount to 2 decimal places
groupedAges = groupedAges.round(2)

# Display Grouped Ages Table
groupedAges.sort_index()

Total Count  Percentage of Players
<10             23                   2.95
10-14           28                   3.59
15-19          136                  17.44
20-24          365                  46.79
25-29          101                  12.95
30-34           73                   9.36
35-39           41                   5.26
40+             13                   1.67

## Purchasing Analysis (Total)
* Run basic calculations to obtain number of unique items, average price, etc.
* Create a summary data frame to hold the results
* Give the displayed data cleaner formatting
* Display the summary data frame

In [14]:
purchase_data.describe()

Purchase ID         Age     Item ID       Price
count   780.000000  780.000000  780.000000  780.000000
mean    389.500000   22.714103   92.114103    3.050987
std     225.310896    6.659444   52.775943    1.169549
min       0.000000    7.000000    0.000000    1.000000
25%     194.750000   20.000000   48.000000    1.980000
50%     389.500000   22.000000   93.000000    3.150000
75%     584.250000   25.000000  139.000000    4.080000
max     779.000000   45.000000  183.000000    4.990000

In [15]:
# listUniqueItems = purchase_data["Item ID"].value_counts() # just displays the list of items and how
# print(listUniqueItems)many times they appear
uniqueItems = len(purchase_data["Item ID"].value_counts())
# print(uniqueItems)
numberPurchases = len(purchase_data["Price"])
# print (numberPurchases)
averageItemPrice = round(purchase_data["Price"].mean(), 2)
# print (averageItemPrice)
totalRevenue = purchase_data["Price"].sum()
# print (f"${totalRevenue}")

# Insert values into dataframe using dictionary of list style dataframe creation
summary_table = pd.DataFrame({"Number of Unique Items": [uniqueItems],
                              "Number of Purchases": [numberPurchases],
                              "Average Price": [averageItemPrice],
                              "Total Revenue": [totalRevenue]})
summary_table

Number of Unique Items  Number of Purchases  Average Price  Total Revenue
0                     183                  780           3.05        2379.77

## Purchasing Analysis (Gender)
* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person by gender
* Create a summary data frame to hold the results
* Give the displayed data cleaner formatting
* Display the summary data frame

In [16]:
# Run basic calculations
gender_purchase_total = purchase_data.groupby(["Gender"]).sum()["Price"].rename("Total Purchase Value") # list of floats
average_by_gender = purchase_data.groupby(["Gender"]).mean()["Price"].rename("Average Purchase Price") # list of floats
gender_counts = purchase_data.groupby(["Gender"]).count()["Price"].rename("Purchase Count") # list of floats
print(gender_purchase_total)
print("*"*50)
print(average_by_gender)
print("*"*50)
print(gender_counts)
print("*"*50)

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Total Purchase Value, dtype: float64
**************************************************
Gender
Female                   3.203009
Male                     3.017853
Other / Non-Disclosed    3.346000
Name: Average Purchase Price, dtype: float64
**************************************************
Gender
Female                   113
Male                     652
Other / Non-Disclosed     15
Name: Purchase Count, dtype: int64
**************************************************


In [17]:
# Convert to DataFrame
gender_data = pd.DataFrame({"Purchase Count": gender_counts, "Average Purchase Price": average_by_gender, "Total Purchase Value": gender_purchase_total})
gender_data

Purchase Count  Average Purchase Price  \
Gender                                                          
Female                            113                3.203009   
Male                              652                3.017853   
Other / Non-Disclosed              15                3.346000   

                       Total Purchase Value  
Gender                                       
Female                               361.94  
Male                                1967.64  
Other / Non-Disclosed                 50.19

used https://mkaz.blog/code/python-string-format-cookbook/ for formmatting dollar signs

In [18]:
gender_data["Average Purchase Price"] = gender_data["Average Purchase Price"].map("$ {:,.2f}".format)
gender_data["Total Purchase Value"] = gender_data["Total Purchase Value"].map("$ {:,.2f}".format)
gender_data ["Purchase Count"] = gender_data["Purchase Count"].map("{:,}".format)

gender_data = gender_data.loc[:, ["Purchase Count", "Average Purchase Price", "Total Purchase Value"]]

# Display the Gender Table
gender_data

Purchase Count Average Purchase Price  \
Gender                                                        
Female                           113                 $ 3.20   
Male                             652                 $ 3.02   
Other / Non-Disclosed             15                 $ 3.35   

                      Total Purchase Value  
Gender                                      
Female                            $ 361.94  
Male                            $ 1,967.64  
Other / Non-Disclosed              $ 50.19

## Purchasing Analysis (Age)
* Bin the purchase_data data frame by age
* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
* Create a summary data frame to hold the results
* Give the displayed data cleaner formatting
* Display the summary data frame

In [19]:
# Bin the Purchasing Data
purchase_data["Age Ranges"] = pd.cut(purchase_data["Age"], age_bins, labels=group_names)

# Run basic calculations
purchase_total_by_age = purchase_data.groupby(["Age Ranges"]).sum()["Price"].rename("Total Purchase Value")
average_by_age = purchase_data.groupby(["Age Ranges"]).mean()["Price"].rename("Average Purchase Price")
counts_by_age = purchase_data.groupby(["Age Ranges"]).count()["Price"].rename("Purchase Count")

# Convert to DataFrame
purchases_by_age_df = pd.DataFrame({"Purchase Count": counts_by_age, "Average Purchase Price": average_by_age, "Total Purchase Value": purchase_total_by_age})

# Give the displayed data cleaner formatting by rounding dollar amount to 2 decimal places
purchases_by_age_df["Total Purchase Value"] = purchases_by_age_df["Total Purchase Value"].map("${:,.2f}".format)
purchases_by_age_df["Average Purchase Price"] = purchases_by_age_df["Average Purchase Price"].map("${:,.2f}".format)
purchases_by_age_df["Purchase Count"] = purchases_by_age_df["Purchase Count"].map("{:,}".format)
purchases_by_age_df = purchases_by_age_df.loc[:, ["Purchase Count", "Average Purchase Price", "Total Purchase Value"]]

# Display the summary data frame
purchases_by_age_df

Purchase Count Average Purchase Price Total Purchase Value
Age Ranges                                                           
<10                    23                  $3.35               $77.13
10-14                  28                  $2.96               $82.78
15-19                 136                  $3.04              $412.89
20-24                 365                  $3.05            $1,114.06
25-29                 101                  $2.90              $293.00
30-34                  73                  $2.93              $214.00
35-39                  41                  $3.60              $147.67
40+                    13                  $2.94               $38.24

## Top Spenders
* Run basic calculations to obtain the results in the table below
* Create a summary data frame to hold the results
* Sort the total purchase value column in descending order
* Give the displayed data cleaner formatting
* Display a preview of the summary data frame

In [20]:
# Run basic calculations to obtain the results in the table below
user_count = purchase_data.groupby(["SN"]).count()["Price"].rename("Purchase Count")
user_average = purchase_data.groupby(["SN"]).mean()["Price"].rename("Average Purchase Price")
user_total = purchase_data.groupby(["SN"]).sum()["Price"].rename("Total Purchase Value")

# Create a summary data frame to hold the results
user_data_df = pd.DataFrame({"Total Purchase Value": user_total, "Average Purchase Price": user_average, "Purchase Count": user_count})

# Sort the total purchase value column in descending order
user_sorted = user_data_df.sort_values("Total Purchase Value", ascending=False)

# Give the displayed data cleaner formatting by rounding dollar amount to 2 decimal places
user_sorted = user_sorted.loc[:,["Purchase Count", "Average Purchase Price", "Total Purchase Value"]]
user_sorted["Total Purchase Value"] = user_sorted["Total Purchase Value"].map("${:,.2f}".format)
user_sorted["Average Purchase Price"] = user_sorted["Average Purchase Price"].map("${:,.2f}".format)

# Display a preview of the summary data frame
user_sorted.head(5)

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items
* Retrieve the Item ID, Item Name, and Item Price columns
* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value
* Create a summary data frame to hold the results
* Sort the purchase count column in descending order
* Give the displayed data cleaner formatting
* Display a preview of the summary data frame

In [21]:
# Retrieve the Item ID, Item Name, and Item Price columns
item_data = purchase_data.loc[:,["Item ID", "Item Name", "Price"]]

# Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value
item_count = item_data.groupby(["Item ID", "Item Name"]).count()["Price"].rename("Purchase Count")
average_item_purchase = item_data.groupby(["Item ID", "Item Name"]).mean()["Price"]
total_item_purchase = item_data.groupby(["Item ID", "Item Name"]).sum()["Price"].rename("Total Purchase Value")

# Create a summary data frame to hold the results
item_data_pd = pd.DataFrame({"Total Purchase Value": total_item_purchase, "Item Price": average_item_purchase, "Purchase Count": item_count})

# Sort the purchase count column in descending order
item_data_count_sorted = item_data_pd.sort_values("Purchase Count", ascending=False)

# Give the displayed data cleaner formatting by rounding dollar amount to 2 decimal places
item_data_count_sorted["Purchase Count"] = item_data_count_sorted["Purchase Count"].map("{:,}".format)
item_data_count_sorted["Item Price"] = item_data_count_sorted["Item Price"].map("${:,.2f}".format)
item_data_count_sorted["Total Purchase Value"] = item_data_count_sorted["Total Purchase Value"].map("${:,.2f}".format)
item_popularity = item_data_count_sorted.loc[:,["Purchase Count", "Item Price", "Total Purchase Value"]]

item_popularity.head(5)

Purchase Count  \
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm             12   
145     Fiery Glass Crusader                                      9   
108     Extraction, Quickblade Of Trembling Hands                 9   
82      Nirvana                                                   9   
19      Pursuit, Cudgel of Necromancy                             8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
82      Nirvana                                           $4.90   
19      Pursuit, Cudgel of Necromancy                     $1.02   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
82      Nirvana                                                    $44.10  
19      Pursuit, Cudgel of Necromancy                               $8.16

## Most Profitable Items
* Sort the above table by total purchase value in descending order
* Give the displayed data cleaner formatting
* Display a preview of the data frame

In [22]:
# Item Table (Sorted by Total Purchase Value)
total_purchase_value = item_data_pd.sort_values("Total Purchase Value", ascending=False)

# Give the displayed data cleaner formatting by rounding dollar amount to 2 decimal places
total_purchase_value["Item Price"] = total_purchase_value["Item Price"].map("${:,.2f}".format)
total_purchase_value["Purchase Count"] = total_purchase_value["Purchase Count"].map("{:,}".format)
total_purchase_value["Total Purchase Value"] = total_purchase_value["Total Purchase Value"].map("${:,.2f}".format)
profitable_items = total_purchase_value.loc[:,["Purchase Count", "Item Price", "Total Purchase Value"]]

# Display a preview of the data frame
profitable_items.head(5)

Purchase Count  \
Item ID Item Name                                                     
178     Oathbreaker, Last Hope of the Breaking Storm             12   
82      Nirvana                                                   9   
145     Fiery Glass Crusader                                      9   
92      Final Critic                                              8   
103     Singed Scalpel                                            8   

                                                     Item Price  \
Item ID Item Name                                                 
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
92      Final Critic                                      $4.88   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
92      Final Critic                                               $39.04  
103     Singed Scalpel                                             $34.80